### PR TITLE
Basic String Functionality

### PR DESCRIPTION
**Pull request by BaldPulse**

### CODE REVIEW

File: compiler.ts [hunk 1]
```diff
@@ -195,6 +197,30 @@ function codeGenValue(val: Value<[Type, SourceLocation]>, env: GlobalEnv): Array
       return ["(i32.const " + val.value + ")"];
     case "bool":
       return [`(i32.const ${Number(val.value)})`];
+    // case "str":
```
Comment (BaldPulse) [inline compiler.ts]: Clean up

File: ir.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import {Type, BinOp, UniOp, Parameter} from './ast';
 
-export type Program<A> = { a?: A, funs: Array<FunDef<A>>, inits: Array<VarInit<A>>, classes: Array<Class<A>>, body: Array<BasicBlock<A>> }
+export type Program<A> = { a?: A, funs: Array<FunDef<A>>, inits: Array<VarInit<A>>,  classes: Array<Class<A>>, body: Array<BasicBlock<A>> }
```
Comment (BaldPulse) [inline ir.ts]: delete this space

File: lower.ts [hunk 3]
```diff
@@ -25,6 +26,7 @@ export function lowerProgram(p : AST.Program<[Type, SourceLocation]>, env : Glob
     var blocks : Array<IR.BasicBlock<[Type, SourceLocation]>> = [];
     var firstBlock : IR.BasicBlock<[Type, SourceLocation]> = {  a: p.a, label: generateName("$startProg"), stmts: [] }
     blocks.push(firstBlock);
+    // var strInit = initStrings(p.inits);
```
Comment (BaldPulse) [inline lower.ts:29]: clean up

File: runner.ts [hunk 4]
```diff
@@ -68,6 +68,7 @@ export function augmentEnv(env: GlobalEnv, prog: Program<[Type, SourceLocation]>
 
 // export async function run(source : string, config: Config) : Promise<[Value, compiler.GlobalEnv, GlobalTypeEnv, string]> {
 export async function run(source : string, config: Config) : Promise<[Value, GlobalEnv, GlobalTypeEnv, string, WebAssembly.WebAssemblyInstantiatedSource]> {
+  // console.log(JSON.stringify(config.importObject));
```
Comment (BaldPulse) [inline runner.ts]: clean up

File: tests/import-object.test.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
 import { readFileSync } from "fs";
+import { CLASS } from "../utils";
```
Comment (BaldPulse) [inline tests/import-object.test.ts]: Make sure this file has no changes from the origin

File: parser.ts [hunk 6]
```diff
@@ -136,12 +141,25 @@ export function traverseExpr(c : TreeCursor, s : string) : Expr<SourceLocation>
         case "<":
           op = BinOp.Lt;
           break;
+        // c.nextSibling(); // go to rhs
```
Comment (BaldPulse) [inline parser.ts]: clean up

File: type-check.ts [hunk 7]
```diff
@@ -31,15 +31,15 @@ export const defaultTypeEnv = {
   classes: new Map(),
 };
 
-export function emptyGlobalTypeEnv() : GlobalTypeEnv {
+export function emptyGlobalTypeEnv(): GlobalTypeEnv {
```
Comment (BaldPulse) [inline type-check.ts]: fix the spaces

File: webstart.ts [hunk 8]
```diff
@@ -1,7 +1,8 @@
 import {BasicREPL} from './repl';
 import { Type, Value } from './ast';
 import { defaultTypeEnv } from './type-check';
-import { NUM, BOOL, NONE } from './utils';
+import { NUM, BOOL, NONE, CLASS } from './utils';
+import { type } from 'os';
```
Comment (BaldPulse) [inline webstart.ts]: delete this line

File: webstart.ts [hunk 9]
```diff
@@ -48,6 +68,8 @@ function webStart() {
         print_num: (arg: number) => print(NUM, arg),
         print_bool: (arg: number) => print(BOOL, arg),
         print_none: (arg: number) => print(NONE, arg),
+        print_str: (arg: number) => print(CLASS("str"), arg, memory),
+        // print_str: (arg: number) => print(NUM, arg),
```
Comment (BaldPulse) [inline webstart.ts]: clean up

**Review by BaldPulse — COMMENTED** [review]
Pending fixes

File: runner.ts [hunk 10]
```diff
@@ -105,6 +106,7 @@ export async function run(source : string, config: Config) : Promise<[Value, Glo
     (func $print_num (import "imports" "print_num") (param i32) (result i32))
     (func $print_bool (import "imports" "print_bool") (param i32) (result i32))
     (func $print_none (import "imports" "print_none") (param i32) (result i32))
+    (func $print_str (import "imports" "print_str") (param i32) (result i32))
```
Comment (shupohan) [inline runner.ts:108]: add print_str function for printing the string

File: ast.ts [hunk 11]
```diff
@@ -41,7 +41,7 @@ export type Expr<A> =
   | {  a?: A, tag: "lookup", obj: Expr<A>, field: string }
   | {  a?: A, tag: "index", obj: Expr<A>, index: Expr<A> }
   | {  a?: A, tag: "method-call", obj: Expr<A>, method: string, arguments: Array<Expr<A>> }
-  | {  a?: A, tag: "construct", name: string }
+  | {  a?: A, tag: "construct", name: string , strarg?:string}
```
Comment (shupohan) [inline ast.ts:44]: add a annotation for method

File: lower.ts [hunk 12]
```diff
@@ -264,6 +320,9 @@ function flattenExprToExpr(e : AST.Expr<[Type, SourceLocation]>, env : GlobalEnv
         offset: { tag: "wasmint", value: offset }}];
     }
     case "construct":
+      if(e.name == "str"){
+        return lowerStr({tag:"str", value:e.strarg}, e.a[1]);
```
Comment (shupohan) [inline lower.ts:324]: adding lowerStr in the case "construct"

File: lower.ts [hunk 13]
```diff
@@ -81,6 +83,60 @@ function literalToVal(lit: AST.Literal) : IR.Value<[Type, SourceLocation]> {
...
+  const alloc : IR.Expr<[Type, SourceLocation]> = { tag: "alloc", amount: { tag: "wasmint", value: Math.ceil(lit.value.length / 4) + 1 } };
+  const assigns : IR.Stmt<[Type, SourceLocation]>[] = [];
+  assigns.push({
+    tag: "store",
+    start: { tag: "id", name: strName },
+    offset: { tag: "wasmint", value: 0 },
+    value: { tag: "wasmint", value: lit.value.length }
+  });
+
+  // var result = ( ( (bytes[0] & 0xFF) << 8) | (bytes[1] & 0xFF) ); charCodeAt(i)
```
Comment (shupohan) [inline lower.ts:97]: write one byte for each character

File: string.wat [hunk 14]
```diff
@@ -0,0 +1,43 @@
+(func $str$access (param $self i32) (param $index i32) (result i32))
+(func $str$length (param $self i32) (param $index i32) (result i32))
```
Comment (shupohan) [inline string.wat:2]: add $str$length function for len()

File: string.wat [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    (local.get $self)
+    (call $assert_not_none)
+    (i32.const 0)
+    (call $load) ;; load the length of the string
+    (return)
+(func $str$lessthan (param $self i32) (param $rhs i32) (result i32))
+(func $str$greaterthan (param $self i32) (param $rhs i32) (result i32))
+(func $str$equalsto (param $self i32) (param $rhs i32) (result i32))
```
Comment (shupohan) [inline string.wat:10]: add $str$equalsto for string comparison

File: webstart.ts [hunk 16]
```diff
@@ -16,11 +17,30 @@ function stringify(typ: Type, arg: any) : string {
   }
 }
 
-function print(typ: Type, arg : number) : any {
+function print(typ: Type, arg : number, mem?: WebAssembly.Memory) : any {
   console.log("Logging from WASM: ", arg);
-  const elt = document.createElement("pre");
-  document.getElementById("output").appendChild(elt);
-  elt.innerText = stringify(typ, arg);
+  if(typ.tag === "class"){
```
Comment (shupohan) [inline webstart.ts:21]: printing string for print_str

**Review by shupohan — APPROVED** [review]
Briefly look through the main files, the testing files do not look carefully

**Pull request merged**